src/librustc/ty/layout.rs

```diff
@@ -606,34 +609,219 @@ impl<'a, 'tcx> LayoutCx<'tcx, TyCtxt<'a, 'tcx, 'tcx>> {
                     &ReprOptions::default(),
                     StructKind::AlwaysSized)?;
+                let (prefix_size, prefix_align) = (prefix.size, prefix.align);
+
+                // Split the prefix layout into the "outer" fields (upvars and
+                // discriminant) and the "promoted" fields. Promoted fields will
+                // get included in each variant that requested them in
+                // GeneratorLayout.
+                let renumber_indices = |mut index: Vec<u32>| -> Vec<u32> {
+                    debug!("renumber_indices({:?})", index);
+                    let mut inverse_index = (0..index.len() as u32).collect::<Vec<_>>();
+                    inverse_index.sort_unstable_by_key(|i| index[*i as usize]);
+                    for i in 0..index.len() {
+                        index[inverse_index[i] as usize] = i as u32;
+                    }
+                    debug!("renumber_indices() => {:?}", index);
+                    index
+                };
+                debug!("prefix = {:#?}", prefix);
+                let (outer_fields, promoted_offsets, promoted_memory_index) = match prefix.fields {
+                    FieldPlacement::Arbitrary { offsets, memory_index } => {
+                        let (offsets_a, offsets_b) =
+                            offsets.split_at(discr_index + 1);
+                        let (memory_index_a, memory_index_b) =
+                            memory_index.split_at(discr_index + 1);
+                        let outer_fields = FieldPlacement::Arbitrary {
+                            offsets: offsets_a.to_vec(),
+                            memory_index: renumber_indices(memory_index_a.to_vec())
+                        };
+                        (outer_fields,
+                         offsets_b.to_vec(),
+                         renumber_indices(memory_index_b.to_vec()))
+                    }
+                    _ => bug!(),
+                };
 
                 let mut size = prefix.size;
                 let mut align = prefix.align;
-                let variants_tys = substs.state_tys(def_id, tcx);
-                let variants = variants_tys.enumerate().map(|(i, variant_tys)| {
+                let variants = info.variant_fields.iter_enumerated().map(|(index, variant_fields)| {
+                    // Only include overlap-eligible fields when we compute our variant layout.
+                    let variant_only_tys = variant_fields.iter().flat_map(|local| {
+                        let ty = info.field_tys[*local];
+                        match assignments[*local] {
+                            Unassigned => bug!(),
+                            Assigned(v) if v == index => Some(subst_field(ty)),
+                            Assigned(_) => bug!("assignment does not match variant"),
+                            Ineligible(_) => None,
+                        }
+                    });
```

Comment:
This is using `flat_map` when `filter_map` would succeed - but also, it looks like `filter` *then* `map` would still work - IMO, it would make `subst_field(info.field_tys[local])` clearer, while also simplifying the `filter` condition.